sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/HostnameRule.java

```diff
@@ -20,85 +20,109 @@
      *
-     * @param hostname hostname of the mesos agent to require
+     * @param matcher matcher for hostname of the mesos agent to require
      */
-    public static PlacementRule require(String hostname) {
-        return new HostnameRule(hostname);
+    public static PlacementRule require(StringMatcher matcher) {
+        return new HostnameRule(matcher);
     }
 
     /**
      * Requires that a task be placed on one of the provided hostnames.
      *
-     * @param hostnames hostnames of the mesos agents to require
+     * @param matchers matchers for hostnames of the mesos agents to require
      */
-    public static PlacementRule require(Collection<String> hostnames) {
-        return new OrRule(toHostnameRules(hostnames));
+    public static PlacementRule require(Collection<StringMatcher> matchers) {
+        if (matchers.size() == 1) {
+            return require(matchers.iterator().next());
+        }
+        return new OrRule(toHostnameRules(matchers));
     }
 
     /**
      * Requires that a task be placed on one of the provided hostnames.
      *
-     * @param hostnames hostnames of the mesos agents to require
+     * @param matchers matchers for hostnames of the mesos agents to require
      */
-    public static PlacementRule require(String... hostnames) {
-        return require(Arrays.asList(hostnames));
+    public static PlacementRule require(StringMatcher... matchers) {
+        return require(Arrays.asList(matchers));
     }
 
     /**
      * Requires that a task NOT be placed on the provided hostname.
      *
-     * @param hostname hostname of the mesos agent to avoid
+     * @param matcher matcher for hostname of the mesos agent to avoid
      */
-    public static PlacementRule avoid(String hostname) {
-        return new NotRule(require(hostname));
+    public static PlacementRule avoid(StringMatcher matcher) {
+        return new NotRule(require(matcher));
     }
 
     /**
      * Requires that a task NOT be placed on any of the provided hostnames.
      *
-     * @param hostnames hostnames of the mesos agents to avoid
+     * @param matchers matchers for hostnames of the mesos agents to avoid
      */
-    public static PlacementRule avoid(Collection<String> hostnames) {
-        return new NotRule(require(hostnames));
+    public static PlacementRule avoid(Collection<StringMatcher> matchers) {
+        if (matchers.size() == 1) {
+            return avoid(matchers.iterator().next());
+        }
+        return new NotRule(require(matchers));
     }
 
     /**
      * Requires that a task NOT be placed on any of the provided hostnames.
      *
-     * @param hostnames hostnames of the mesos agents to avoid
+     * @param matchers matchers for hostnames of the mesos agents to avoid
      */
-    public static PlacementRule avoid(String... hostnames) {
-        return avoid(Arrays.asList(hostnames));
+    public static PlacementRule avoid(StringMatcher... matchers) {
```

Comment:
I had indeed tried to have that but ended up getting type conflicts with java. As a workaround, I added the `toStringMatchers()` functions further up for devs to use like this: `HostnameRule.avoid(toStringMatchers(exactValues))`

An alternate method would be to add new differently-named functions, e.g. `avoidExact()`/`requireExact()`, to avoid the type collision issues. Any preference?